compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -596,13 +596,30 @@ class Typer extends Namer
         .withSpan(tree.span)
         .computeNullable()
 
-    def typeSelectOnType(qual: untpd.Tree)(using Context) =
-      typedSelect(untpd.cpy.Select(tree)(qual, tree.name.toTypeName), pt)
+    def javaSelectOnType(qual: Tree)(using Context) =
+      // semantic name conversion for `O$` in java code
+      if !qual.symbol.is(JavaDefined) && tree.name.endsWith("$")
+      then
+        val nameOrig = tree.name.asSimpleName
+        val nameSemantic = nameOrig.slice(0, nameOrig.length - 1).moduleClassName
+        val tree2 = untpd.cpy.Select(tree)(qual, nameSemantic)
```

Comment:
unmangleClassName in NameOps already has similar logic

Reply:
Thanks for the tip 👍